### PR TITLE
create_html_table

### PR DESCRIPTION
**Pull request by yangchristina**

### CODE REVIEW

**Comment by yangchristina** [comment]
@firasm Is there a way to render html from server.py? Because I am getting this in prairelearn:
Screenshot 2023-08-16 at 11 51 43 AM

with this in markdown:
```md
Consider the following PMF.

{{ params.table }}

Where appropriate round answers to 3 decimal places.
```

**Comment by firasm** [comment]
yes, use triple brackets:

```{{{ params.table }}}```

File: src/problem_bank_helpers/problem_bank_helpers.py [hunk 1]
```diff
@@ -382,11 +382,49 @@ def base64_decode_file(file):
...
+    """Encode a string to base64 and add it as the user submitted code file
     """
     # partially based off of https://github.com/PrairieLearn/PrairieLearn/blob/2ff7c5cc2435bae80c0ba512631749f9c3eadb43/apps/prairielearn/elements/pl-file-upload/pl-file-upload.py#L114C1-L119
     parsed_file = {"name": "user_code.py", "contents": base64_encode_string(answer)}
     if isinstance(data["submitted_answers"].get("_files", None), list):
         data["submitted_answers"]["_files"].append(parsed_file)
     else:
         data["submitted_answers"]["_files"] = [parsed_file]
+
+def convert_markdown_table(table, width="100%", first_row_is_header=True, first_col_is_header=True, wrap_latex=False, wrap_header_latex=False):
```
Comment (firasm) [inline src/problem_bank_helpers/problem_bank_helpers.py]: I would change the name of this function and rename the `wrap_latex` argument, and reorder the last two arguments.

```suggestion
def create_html_table(table, width="100%", first_row_is_header=True, first_col_is_header=True, wrap_header_latex=False, wrap_nonheader_latex=False, ):
```

File: src/problem_bank_helpers/problem_bank_helpers.py [hunk 2]
```diff
@@ -382,11 +382,49 @@ def base64_decode_file(file):
...
+    Args:
+        table (list): A list of lists representing the table
+        width (str, optional): The width of the table. Ex. "100%", "500px", etc.
+        first_row_is_header (bool, optional): Whether the first row is a header. Defaults to True.
+        first_col_is_header (bool, optional): Whether the first column is a header. Defaults to True.
+        wrap_latex (bool, optional): Whether to wrap all non-header table cells in $ for LaTeX. Defaults to False.
+        wrap_header_latex (bool, optional): Whether to wrap all header table cells in $ for LaTeX. Defaults to False.
+    """
+    def wrap(x):
+        return f"${x}$" if wrap_latex else x
```
Comment (firasm) [inline src/problem_bank_helpers/problem_bank_helpers.py]:
```suggestion
        return f"${x}$" if wrap_nonheader_latex else x
```

File: src/problem_bank_helpers/problem_bank_helpers.py [hunk 3]
```diff
@@ -382,11 +382,49 @@ def base64_decode_file(file):
...
+    Example usage:\n
+    server.py: data["params"]["table1"] = pbh.convert_markdown_table([["a", "b", "c"], ["x", "1"]], wrap_latex=True)\n
+    markdown: {{{ params.table1 }}}
+
+    Args:
+        table (list): A list of lists representing the table
+        width (str, optional): The width of the table. Ex. "100%", "500px", etc.
+        first_row_is_header (bool, optional): Whether the first row is a header. Defaults to True.
+        first_col_is_header (bool, optional): Whether the first column is a header. Defaults to True.
+        wrap_latex (bool, optional): Whether to wrap all non-header table cells in $ for LaTeX. Defaults to False.
```
Comment (firasm) [inline src/problem_bank_helpers/problem_bank_helpers.py]:
```suggestion
        wrap_nonheader_latex (bool, optional): Whether to wrap all non-header table cells in $ for LaTeX. Defaults to False.
```

File: src/problem_bank_helpers/problem_bank_helpers.py [hunk 4]
```diff
@@ -382,11 +382,49 @@ def base64_decode_file(file):
...
     if isinstance(data["submitted_answers"].get("_files", None), list):
         data["submitted_answers"]["_files"].append(parsed_file)
     else:
         data["submitted_answers"]["_files"] = [parsed_file]
+
+def convert_markdown_table(table, width="100%", first_row_is_header=True, first_col_is_header=True, wrap_latex=False, wrap_header_latex=False):
+    """
+    Convert a python table to HTML\n
+    Example usage:\n
+    server.py: data["params"]["table1"] = pbh.convert_markdown_table([["a", "b", "c"], ["x", "1"]], wrap_latex=True)\n
```
Comment (firasm) [inline src/problem_bank_helpers/problem_bank_helpers.py]:
```suggestion
    server.py: data["params"]["table1"] = pbh.convert_markdown_table([["a", "b", "c"], ["x", "1"]], wrap_nonheader_latex=True)\n
```

**Review by firasm — COMMENTED** [review]
I'm not sure about the `wrap_header_latex` and `wrap_nonheader_latex` ; I almost want that to be the responsibility of the user rather than them having be automatically added.

If you think it will save time and reduce headache, then I suggest leaving them in (e.g for long tables it would be cumbersome to add them in manually).

**Comment by yangchristina** [comment]
I don't know if the latex stuff would be useful for other people (so I set default to False), but it would certainly save me time, since all of my tables consist of numbers.

**Pull request merged**